### PR TITLE
Arnej/keep and follow refs

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@geirst @toregge @havardpe please review
I am worried that there are still possible race conditions here, so we need to
discuss this in more detail.

### CODE REVIEW

**Review by geirst — APPROVED** [review]
👍

**Comment by arnej27959** [comment]
@geirst please take a quick look at the new stress test, and see if it follows the expected usage pattern.  It found a few race conditions immediately, and it seems clear I need to keep more data in PrepareResult so we can check if they are still valid in complete_add_document.

File: searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+        auto task = std::make_unique<TwoPhaseRemoveTask>(*this, docid, guard);
+        auto r = multi_prepare_workers.execute(std::move(task));
+        ASSERT_EQ(bool(r), false);
+    }
+    void update_document(uint32_t docid) {
+        size_t vec_num = get_rnd(loaded_vectors.size());
+        ConstVectorRef vec = loaded_vectors[vec_num];
+        auto guard = take_read_guard();
+        auto task = std::make_unique<TwoPhaseUpdateTask>(*this, docid, vec, guard);
+        auto r = multi_prepare_workers.execute(std::move(task));
```
Comment (geirst) [inline searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp:308]: Same comment as with add_document().

File: searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+        ASSERT_EQ(bool(r), false);
+    }
+    void commit(uint32_t docid) {
+        index->transfer_hold_lists(gen_handler.getCurrentGeneration());
+        gen_handler.incGeneration();
+        gen_handler.updateFirstUsedGeneration();
+        index->trim_hold_lists(gen_handler.getFirstUsedGeneration());
+        in_progress->clearBit(docid);
```
Comment (geirst) [inline searchlib/src/tests/tensor/hnsw_index/stress_hnsw_mt.cpp:316]: Is this racing with what happens in gen_operation(). Should we protect it with a mutex?

File: searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp [hunk 3]
```diff
@@ -322,14 +322,21 @@ HnswIndex::internal_prepare_add(uint32_t docid, TypedCells input_vector) const
 }
 
 HnswIndex::LinkArray 
-HnswIndex::filter_valid_docids(const LinkArrayRef &docids)
+HnswIndex::filter_valid_docids(uint32_t level, const LinkArrayRef &docids, uint32_t me)
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/tensor/hnsw_index.cpp]: Consider rename 'me' -> 'me_docid' / 'self_docid'

**Review by geirst — APPROVED** [review]
Looks good to me 👍

**Comment by geirst** [comment]
I have reviewed up to and including "_print more statistics_". Looks good to me 👍

**Pull request merged**